File: src/main/java/leetcode/lists/hot200/LeetCode251.java
package leetcode.lists.hot200;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/3/8 18:44
 */
public class LeetCode251 {

    class Vector2D {
        int row;
        int col;
        int[][] vec;

        public Vector2D(int[][] vec) {
            this.vec = vec;
        }

        public int next() {
            if (!this.hasNext()) {
                return -1;
            }
            int nextRow = row;
            if (col == this.vec[row].length) {
                while (nextRow < this.vec.length) {
                    nextRow++;
                    if (nextRow < this.vec.length && this.vec[nextRow].length > 0) {
                        break;
                    }
                }
                col = 0;
            }

            row = nextRow;
            int val = vec[row][col];
            col++;
            return val;
        }

        public boolean hasNext() {
            if (row == this.vec.length) {
                return false;
            }
            int nextRow = row;
            if (col == this.vec[row].length) {
                while (nextRow < this.vec.length) {
                    nextRow++;
                    if (nextRow < this.vec.length && this.vec[nextRow].length > 0) {
                        break;
                    }
                }
            }
            if (nextRow == this.vec.length) {
                return false;
            }
            return true;
        }
    }
}

File: go/src/leetcode/topinterview/LeetCode251.go
package topinterview

type Vector2D struct {
	vector [][]int
	i      int
	j      int
}

func ConstructorVector2D(vec [][]int) Vector2D {
	return Vector2D{
		vector: vec,
		i:      0,
		j:      0,
	}
}

func (this *Vector2D) Next() int {
	if this.j > len(this.vector[this.i])-1 {
		this.i++
		for this.i < len(this.vector) && len(this.vector[this.i]) == 0 {
			this.i++
		}
		this.j = 0
	}
	num := this.vector[this.i][this.j]
	this.j++
	return num
}

func (this *Vector2D) HasNext() bool {
	row := this.i
	col := this.j
	if len(this.vector) == 0 {
		return false
	}
	if len(this.vector[row]) > 0 && col < len(this.vector[row])-1 {
		return true
	}
	if col > len(this.vector[row])-1 {
		row++
		for row < len(this.vector) && len(this.vector[row]) == 0 {
			row++
		}
		col = 0
	}
	if row == len(this.vector) {
		return false
	}
	return true
}
